pkg/workspace/assert.go

```diff
@@ -0,0 +1,162 @@
+// Copyright 2022 VMware, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package workspace
+
+import (
+	"fmt"
+	"github.com/k14s/starlark-go/starlark"
+	"github.com/vmware-tanzu/carvel-ytt/pkg/filepos"
+	"github.com/vmware-tanzu/carvel-ytt/pkg/template"
+	"github.com/vmware-tanzu/carvel-ytt/pkg/yamlmeta"
+	"github.com/vmware-tanzu/carvel-ytt/pkg/yamltemplate"
+)
+
+// Declare @assert/... annotation names
+const (
+	AnnotationAssertValidate template.AnnotationName = "assert/validate"
+)
+
+// ProcessAndRunValidations takes a root Node and traverses the tree checking for assert annotations.
+// Validations are processed and executed using the value of the annotated node as the parameter to the assertions.
+//
+// When the assertions have violations, the errors are collected and stored in the checker.
+// Otherwise, returns nil.
+func ProcessAndRunValidations(n yamlmeta.Node, checker yamlmeta.Visitor) error {
+	if n == nil {
+		return nil
+	}
+
+	err := yamlmeta.Walk(n, checker)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type assertChecker struct {
+	thread     *starlark.Thread
+	violations []error
+}
+
+func newAssertChecker(threadName string, _ *TemplateLoader) *assertChecker {
+	return &assertChecker{thread: &starlark.Thread{Name: threadName}, violations: []error{}}
+}
+
+func (a *assertChecker) hasViolations() bool {
+	return len(a.violations) > 0
+}
+
+// Visit if `node` is annotated with `@assert/validate` (AnnotationAssertValidate).
+// Checks, validates, and runs the validation Rules, any violations from running the assertions are collected.
+//
+// This visitor returns and error if any assertion is not well-formed,
+// otherwise, returns nil.
+func (a *assertChecker) Visit(node yamlmeta.Node) error {
+	nodeAnnotations := template.NewAnnotations(node)
+	if nodeAnnotations.Has(AnnotationAssertValidate) {
+		switch node.(type) {
+		case *yamlmeta.DocumentSet, *yamlmeta.Array, *yamlmeta.Map:
+			return fmt.Errorf("Invalid @%s annotation - not supported on %s at %s", AnnotationAssertValidate, yamlmeta.TypeName(node), node.GetPosition().AsCompactString())
+		default:
+			rules, syntaxErr := newRulesFromAssertValidateAnnotation(nodeAnnotations[AnnotationAssertValidate], node)
```

Comment:
I dig what you did there: conveying that the error here would be a syntax problem by naming the `err` to `syntaxErr`.

It broke up the monotony for me of `if err != nil {...}`.

Got me thinking about the pros and cons. 🤔   Great experimentation/creativity, here. 💡